<a href="https://colab.research.google.com/github/RMV449/INTELIGENCIA-ARTIFICIAL/blob/main/SHOPPING_REY.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **INTELIGENCIA ARTIFICIAL **
---
# **MONZÓN VENTURA REY MARIO**
---
---
#Práctica de Laboratorio 2:SHOPPING
#Descripción del Problema

Cuando los usuarios compran online, no todos acabarán comprando algo. De
hecho, la mayoría de los visitantes de un sitio web de compras en línea
probablemente no terminen realizando una compra durante esa sesión de
navegación web. Sin embargo, podría ser útil que un sitio web de compras
pueda predecir si un usuario tiene la intención de realizar una compra o no: tal
vez mostrar contenido diferente al usuario, como mostrarle una oferta de
descuento si el sitio web cree que el usuario no planea completar la compra.
¿Cómo podría un sitio web determinar la intención de compra de un
usuario? Ahí es donde entrará en juego el aprendizaje automático (Machine
Learning).

#Objetivo
Construir un clasificador del vecino más cercano (k- nearest-neighbor
classifier) para resolver el problema shopping descrito.
Dada la información sobre un usuario (cuántas páginas ha visitado, si está
comprando en un fin de semana, qué navegador web está usando, etc.), el
clasificador predecirá si el usuario realizará una compra o no.

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, classification_report

## Dataset

Se trabajará con un dataset de un sitio web de compras de aproximadamente
12.000 sesiones de usuarios representadas como una fila para cada sesión de
usuario.

In [ ]:
csv_file = pd.read_csv("shopping.csv")
data = pd.read_csv('banknotes.csv')
# Separar las etiquetas (Revenue) de la evidencia
labels = csv_file['Revenue']
evidence = csv_file.drop(columns=['Revenue'])

# Convertir las variables categóricas a variables numéricas
evidence = pd.get_dummies(evidence)

# Verificar las primeras filas de etiquetas y evidencia
print(labels[:5])
print(evidence.head())

0    False
1    False
2    False
3    False
4    False
Name: Revenue, dtype: bool
   Administrative  Administrative_Duration  Informational  \
0               0                      0.0              0   
1               0                      0.0              0   
2               0                      0.0              0   
3               0                      0.0              0   
4               0                      0.0              0   

   Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                     0.0               1                 0.000000   
1                     0.0               2                64.000000   
2                     0.0               1                 0.000000   
3                     0.0               2                 2.666667   
4                     0.0              10               627.500000   

   BounceRates  ExitRates  PageValues  SpecialDay  ...  Month_Jul  Month_June  \
0         0.20       0.20         0.0         0.0

In [ ]:
print(data)

      variance  skewness   curtosis  entropy  class
0    -0.895690   3.00250  -3.606700 -3.44570      1
1     3.476900  -0.15314   2.530000  2.44950      0
2     3.910200   6.06500  -2.453400 -0.68234      0
3     0.607310   3.95440  -4.772000 -4.48530      1
4     2.371800   7.49080   0.015989 -1.74140      0
...        ...       ...        ...      ...    ...
1367 -2.967200 -13.28690  13.472700 -2.62710      1
1368  0.318030  -0.99326   1.094700  0.88619      1
1369 -0.025314  -0.17383  -0.113390  1.21980      1
1370 -2.234000  -7.03140   7.493600  0.61334      1
1371  4.728500   2.10650  -0.283050  1.56250      0

[1372 rows x 5 columns]


## Split training and  testing

In [ ]:
X_training, X_testing, y_training, y_testing = train_test_split(
    evidence, labels, test_size=0.4
)

## Modelo

Si tenemos un conjunto de datos de prueba, podríamos ejecutar nuestro
clasificador con los datos y calcular qué proporción del tiempo clasificamos
correctamente la intención del usuario. Esto nos daría un único porcentaje de
precisión. Pero esa cifra podría ser un poco engañosa. Imagínese, por ejemplo,
si alrededor del 15% de todos los usuarios terminan realizando una compra.
Un clasificador que siempre predijo que el usuario no realizaría una compra,
entonces, lo mediríamos con un 85% de precisión: los únicos usuarios que
clasifica incorrectamente son el 15% de los usuarios que sí realizan una
compra. Y aunque una precisión del 85% suena bastante bien, no parece un
clasificador muy útil.

In [ ]:
model = KNeighborsClassifier(n_neighbors=1)
model.fit(X_training, y_training)

# Evaluación del modelo
accuracy = model.score(X_testing, y_testing)
print(f"Accuracy: {accuracy * 100:.2f}")

Accuracy: 83.39


## Predicción

In [ ]:
predictions = model.predict(X_testing)

## Evaluación

In [ ]:
correct = (y_testing == predictions).sum()
incorrect = (y_testing != predictions).sum()
total = len(predictions)

print(f"Resultados del modelo {type(model).__name__}")
print(f"Correctos: {correct}")
print(f"Incorrectos: {incorrect}")
print(f"Accuracy: {100 * correct / total:.2f}%")

Resultados del modelo KNeighborsClassifier
Correctos: 4113
Incorrectos: 819
Accuracy: 83.39%


#Conclusión

El clasificador k-NN implementado en esta práctica ha mostrado ser eficaz para predecir la intención de compra de los usuarios en un sitio de comercio electrónico, logrando una precisión del 83.39%. Este porcentaje indica que el modelo puede ser valioso para detectar usuarios con alta probabilidad de realizar una compra, lo que podría ser utilizado por las plataformas para personalizar la experiencia del cliente y potencialmente aumentar la tasa de conversión a través de ofertas específicas y contenido adaptado. No obstante, es crucial tener en cuenta la sensibilidad y especificidad del modelo para una evaluación más integral de su desempeño, ya que una alta precisión no siempre garantiza un adecuado balance entre la identificación de compradores y no compradores.

#Anexos

https://colab.research.google.com/drive/13R9y4M0bEnZ2Dt-t4DjdwdAQC0W6yz3J?usp=sharing

https://colab.research.google.com/gist/RMV449/68933e1b30642ad9f27070e61754ef93/shopping_rey.ipynb

